README.md
# Server Clock Bookmarklet

**Server Clock** is a JavaScript bookmarklet that attempts to solve the problem of server/client clock differences. By displaying an accurately approximated server time, this tool allows its users to know exactly when to engage in time-sensitive activities such as ticketing and course registrations.  
It is a reimplementation of [the previous project](https://github.com/r8btx/Server-Clock) with some improvements and additional features.

## Disclaimer

While **Server Clock** aims to be accurate, no warranty is given.

## Quick Setup

- The most up-to-date bookmarklets are automatically generated on **[Server Clock bookmarklet distribution page](https://r8btx.github.io/Server-Clock-Bookmarklet/page)**. For a quick setup, visit the page and follow the instructions.

If you wish to generate the bookmarklet manually from the source, you can do so using [this tool](https://chriszarate.github.io/bookmarkleter/).

## How to Use

1. Navigate to your target webpage. Example: [time.is](https://time.is)
2. Once the webpage is fully loaded, click `Server Clock` bookmarklet on your bookmark bar.
3. A small box will appear on your webpage. After the synchronization process, the web server's time will be displayed.

Tip 1: You can right-click on the box to use additional features.  
Tip 2: You can drag the box around for a better view.  
Tip 3: You can open the browser console (accessed via `F12` key) to see logs.  
Tip 4: The clock is most accurate when the clock is freshly synchronized.

## Notes

1. For the changelog, visit the [ChangeLog.md](./changelog.md) file.
2. To test Server Clock's behavior under your network conditions, use test_accuracy.js from the [Server Clock bookmarklet distribution page](https://r8btx.github.io/Server-Clock-Bookmarklet/page).

## Screenshots

<img src="./img/contextmenu.png" width="800" alt="Screenshot of the Server Clock bookmarklet in action."/>

## How It Works (Overview)

1. Record the base time.
2. Send a HTTP GET request to the target web server using fetch().
3. The web server processes the request and sends back a HTTP response along with a generated `date` header.
4. Estimate the time difference between the base time and the time when the `date` header is generated. The estimation is based on [PerformanceObserver](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceObserver).
5. The estimated server clock will be the result of the following calculation:
   ```txt
   Current Client Time + (HTTP `date` - (Base Time + Estimated Time-To-Respond))
   = Current Client Time + (HTTP `date` - Estimated Client Time when `date` is generated)
   = Current Client Time + Estimated Client/Server Time Difference
   = Estimated Current Server Time
   ```
6. Repeat the process and find the most accurate estimation.

## How It Works (Detail)

### Estimating the Time Difference

<img src="./img/estimatedDifference.png" width="900" alt="Estimating the Time Difference"/>
  
The `date` format appears as `Sat, 1 Jan 2000 12:00:00 GMT`, where time units below 1 second are omitted. Although it is not guaranteed, we assume that the omitted time units are truncated rather than rounded. For example, the trailing 0.999sec should be interpreted as 0sec instead of 1sec. Consequently, any accurate estimation of server time will always fall between the values of `date` and `date + 1sec` (represented by red bars).

During the synchronization process, **Server Clock** estimates `Base Time + Estimated Time-To-Respond`, as indicated by the second green bar in the image above. This estimation allows us to determine the `Client/Server Time Difference`.

The `date` value and its actual generation time can have a time difference ranging from 0 to 1 second due to truncation. This discrepancy leads to an inaccuracy (< 1sec) in the estimated `Client/Server Time Difference`. Therefore, it is necessary to implement a method for reducing the estimation inaccuracy.

### Estimation Pooling

<img src="./img/pooling.PNG" width="900" alt="Estimation Pooling"/>

One approach to mitigate the inaccuracy caused by truncation is through estimation pooling. Since server time estimations can be ordered from the most truncated to the least truncated, the rightmost estimation (shown as a red bar in the image above) is most likely to be the most accurate estimation.

To handle outliers, the estimation pooling process only considers estimations within a 1-second time frame that includes the most estimations.

### Improving Accuracy While Minimizing the Number of Required Samples

<img src="./img/modifiedBinarySearch.jpg" width="900" alt="Modified Binary Search"/>

**Server Clock** utilizes a modified binary search algorithm to calculate delays for optimal sample of the target web server's time. With the clockwise distance (the degree of truncation) measured relatively amongst collected samples, an instance of this search process can be shown as follows:

1. Randomly obtain the first sample A. A becomes the reference point.
2. Obtain sample B by targeting A - 500. Since B has a lower clockwise distance, B becomes the reference point.
3. Obtain sample C by targeting B - 250. Since C has a lower clockwise distance, C becomes the reference point.
4. Obtain sample D by targeting C - 125. Since C has a lower clockwise distance, C remains as the reference point.
5. Obtain sample E by targeting C - 62.5. Since C has a lower clockwise distance, C remains as the reference point.
6. Since the difference between C and E are less than the error tolerance (default: 100ms), the algorithm ends.

   Result: C is used as the clock adjustment with an error of 33ms.

Note: The numbers 500, 250, 125, and 62.5 are determined by `1000 / 2^(the number of collected samples)`

### Others

**Server Clock** also attempts to:

1. Utilize a persistent connection to eliminate the overhead of establishing new connections for each request.
2. Prevent the serving of cached content to ensure up-to-date values for the date attribute.
3. Update the resulting clock at the start of each second to accurately synchronize the displayed clock to the estimated server time.
4. Determine if the collected sample size is sufficient to accurately estimate server time.
5. And others.

## Resources and References

- [Server Clock in PowerShell](https://github.com/r8btx/Server-Clock) - My previous project
- [Bookmarkleter](https://chriszarate.github.io/bookmarkleter/) - JavaScript to bookmarklet conversion
- [A Question of Timing](https://blog.cloudflare.com/a-question-of-timing/) - PerformanceObserver Timings Visual
- [Hypertext Transfer Protocol (HTTP/1.1): Semantics and Content](https://httpwg.org/specs/rfc7231.html#header.date) - `date` header in HTTP response
- [Time.is : exact time for any time zone](https://time.is/) - Tested against this web server
- [Time API: Time Zone API](https://www.timeapi.io/) - Tested against this web server. Use [test_accuracy.js](./js/test_accuracy.js) for future testings.


changelog.md
# Changelog

## Server Clock

[v1.4] fixed z-index issue & rewrote clock drag-and-drop  
[v1.3] Rework on multiple instance cases  
[v1.3] Rework on console interface  
[v1.3] Implemented a modified binary search to calculate effective request delays  
[v1.3] Code refactoring  
[v1.2] test_accuracy.js only supports a single instance (accessed via ServerClock)  
[v1.2] Multiple instances of Server Clock is available (accessed via ServerClocks)  
[v1.2] Exit via console is now available  
[v1.1] Warns inaccuracy if pre-defined `validFor` time passes (default: 20mins)  
[v1.1] Users can now interact with the bookmarklet using a browser console (accessed via ServerClock)  
[v1.1] Improvements such as stricter outlier handling, optimized getDelay(), no display delay after initial sync, etc.  
[v1.1] Code refactoring  
[v1.0] Updated UI, including new status message display  
[v1.0] Code rewrite  
[v0.9] Clean up upon exit  
[v0.8] Combined 'Add 5 more samples' and 'Empty collected samples' into 'Rerun' in UI  
[v0.8] Attempt to increase the estimation accuracy with less number of requests.  
[v0.6] Now uses [Performance.now()](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now) to avoid local clock issues  
[v0.5] Added automatic sampling size calculation  
[v0.5] Removed broken timespacing calculation (Now random)  
[v0.4] Added sampling timeout  
[v0.4] Minor adjustments (timespacing, num_repeat, etc.)  
[v0.3] Added context menu UI  
[v0.2] Added UI  
[v0.2] Created [test_accuracy.js](./js/test_accuracy.js) for accuracy testing  
[v0.1] Initial version

## Launcher

[v0.2] Notify policy errors & link to the distribution page  
[v0.1] Created [launcher.js](./js/launcher.js)

## Distribution Page

[v0.6] Code reduction  
[v0.5] Clicking the bookmarklet link will copy the code to the clipboard.  
[v0.4] Now supports Dark theme  
[v0.4] Code rewrite  
[v0.3] Fixed the distribution page not functioning.  
[v0.2] Recreated the bookmarklet distribution page with an experimental feature of automatic bookmarklet generation.  
[v0.1] Created [the bookmarklet page](https://r8btx.github.io/Server-Clock-Bookmarklet/page) to simplify the setup process.


js/test_accuracy.js
const config = {
  sampleMinimum: 3,
  sampleMaximum: 10,
  timeoutAfter: 5000, // In msec. Will retry request after this time
  validFor: 1200, // In sec. Will assume to be inaccurate after x seconds
  errorTolerance: 100,
  targetURL: `https://www.timeapi.io/api/Time/current/zone?timeZone=${
    Intl.DateTimeFormat().resolvedOptions().timeZone
  }`,
};
const defined = typeof window.ServerClock !== 'undefined';
let Time = defined
  ? window.ServerClock.class.Time
  : class Time {
      constructor(SC) {
        this.SC = SC;
        this.config = SC.config;
        this.currentAdjustment = 0; // The chosen adjustment
        this.lastSynchronized = 0;
      }

      static getClientTime() {
        return performance.timeOrigin + performance.now();
      }

      getServerTime() {
        return Time.getClientTime() + this.currentAdjustment;
      }

      synchronize() {
        this.lastSynchronized = performance.now();
        let differenceSamples = []; // A pool of potential clock adjustments (client/server time difference)
        let requestCount = 0;
        let expectedElapseAfterRequest = 0;

        return new Promise((resolve) => {
          const recursiveSynchronization = async (resolver) => {
            try {
              requestCount++;
              const sample = await new Sample(this.SC).collect();
              differenceSamples.push([sample.HTTPAdjustment, sample.estimatedError]);
              differenceSamples.sort((a, b) => b[0] - a[0]);
              this.currentAdjustment = differenceSamples[0][0];
              expectedElapseAfterRequest = sample.elapsedSinceRequest;

              const nextStep = this.determineNextStep(
                requestCount,
                this.currentAdjustment,
                differenceSamples[differenceSamples.length - 1][0],
              );

              switch (nextStep) {
                case 2:
                  differenceSamples = differenceSamples.slice(1, -1);
                case 1:
                  setTimeout(
                    () => recursiveSynchronization(resolver),
                    this.getDelay(expectedElapseAfterRequest, differenceSamples.length),
                  );
                  break;
                case 0:
                  this.chooseAdjustment(differenceSamples);
                  resolver();
              }
            } catch (error) {
              if (requestCount < this.config.sampleMaximum || differenceSamples.length === 0) {
                setTimeout(
                  () => recursiveSynchronization(resolver),
                  this.getDelay(expectedElapseAfterRequest, differenceSamples.length),
                );
              } else {
                console.warn('Inaccuracy Warning: Request limit reached.');
                this.chooseAdjustment(differenceSamples);
              }
            }
          };
          recursiveSynchronization(resolve);
        });
      }

      determineNextStep(requestCount, max, min) {
        if (requestCount < this.config.sampleMinimum) return 1;
        if (requestCount >= this.config.sampleMaximum) {
          console.warn('Inaccuracy Warning: Request limit reached.');
          return 0;
        }
        if (max - min > 1000) return 2; // Likely outlier

        // Return whether the maximum value and the minimum value is around 1 sec (max truncation difference)
        return 1000 - this.config.errorTolerance > max - min ? 1 : 0;
      }

      getDelay(expectedElapseAfterRequest, precisionLevel) {
        const precision = 1000 / Math.pow(2, precisionLevel);
        const sampleOffset = Math.max(precision, this.config.errorTolerance);
        const adjustedTime = Time.getClientTime() + this.currentAdjustment;
        const delay = 1000 - ((adjustedTime + expectedElapseAfterRequest + sampleOffset) % 1000);

        return delay;
      }

      chooseAdjustment(differenceSamples) {
        const candidates = [];
        for (let i = 0; i < differenceSamples.length; i++) {
          for (let j = differenceSamples.length - 1; j >= 0; j--) {
            if (differenceSamples[i][0] - differenceSamples[j][0] < 1000) {
              candidates.push([j - i, i]);
              break;
            }
          }
        }
        candidates.sort((a, b) => b[0] - a[0]);
        this.adjustment = differenceSamples[candidates[0][1]][0];
        const EstimatedError = differenceSamples[candidates[0][1]][1];

        console.log('[Adj]'.padEnd(12, ' ') + '[Err]');
        differenceSamples.forEach((element) => {
          console.log(`${String(element[0]).padEnd(12, ' ')}${element[1]}`);
        });
        console.log(
          'The chosen adjustment is',
          this.adjustment / 1000,
          'sec with an estimated error of',
          EstimatedError,
          'msec.',
        );
        this.lastSynchronized = performance.now();
      }
    };

let Sample = defined
  ? window.ServerClock.class.Sample
  : class Sample {
      constructor(SC) {
        this.SC = SC;
        this.config = SC.config;
        this.timeIdErrorTolerance = 500;
        this.controller = new AbortController();
        this.observer;
        this.requestTimeout;
        this.request = new Request(this.config.targetURL, {
          // Create a request object with no-cache headers
          // This prevents looking up 'date' header from cached responses
          method: 'GET',
          cache: 'no-store',
          headers: new Headers({
            Connection: 'keep-alive',
            'Cache-Control': 'no-cache',
            Pragma: 'no-cache',
            Expires: '0',
          }),
        });
      }

      collect() {
        return new Promise((resolve, reject) => {
          let HTTPTime = 0;
          let API_Time = 0;
          let timeId = 0;
          this.observer = new PerformanceObserver((list) => {
            const entry = list.getEntries().find(({ name }) => name === this.request.url);
            if (entry && Math.abs(timeId - entry.startTime) < this.timeIdErrorTolerance) {
              clearTimeout(this.requestTimeout);
              this.observer.disconnect();
              this.processSample(resolve, entry, requestStartTime, HTTPTime, API_Time);
            } else {
              this.setTimeout(reject);
            }
          });
          this.observer.observe({ type: 'resource' });

          timeId = performance.now();
          const requestStartTime = Time.getClientTime();
          fetch(this.request, { signal: this.controller.signal })
            .then(async (response) => {
              HTTPTime = Date.parse(response.headers.get('date'));
              response.json().then((j) => {
                API_Time = Date.parse(j.dateTime);
              });
            })
            .catch(() => {});

          this.setTimeout(reject);
        });
      }

      processSample(returnSuccess, entry, requestStartTime, HTTPTime, API_Time) {
        const estimatedLatency = (entry.responseStart - entry.requestStart) / 2;
        const elapsedSinceRequest = entry.requestStart - entry.startTime + estimatedLatency;
        const elapsedSinceResponse = entry.responseEnd - entry.responseStart + estimatedLatency;

        // Calculate client/server time difference based on HTTP `date` header
        // Accomodate estimated elapsed time (time taken before server recorded `date`)
        // Accomodate estimated elapsed time for the json response (time passed since the response was made)
        const HTTPAdjustment = HTTPTime - (requestStartTime + elapsedSinceRequest);
        const HTTPEstimation = Time.getClientTime() + HTTPAdjustment;
        const API_Estimation = API_Time + elapsedSinceResponse;
        const estimatedError = HTTPEstimation - API_Estimation;

        const result = {
          elapsedSinceRequest,
          HTTPAdjustment,
          estimatedError,
        };

        console.log('Collected a difference sample.');
        returnSuccess(result);
      }

      setTimeout(returnFail) {
        if (this.requestTimeout) clearTimeout(this.requestTimeout);
        this.requestTimeout = setTimeout(() => {
          console.log('Request timed out. Timeout:', config.timeoutAfter / 1000, 'sec.');
          this.observer.disconnect();
          this.controller.abort();
          returnFail();
        }, this.config.timeoutAfter);
      }
    };

let ServerClock = defined
  ? undefined
  : class ServerClock {
      constructor() {
        this.version = '1.3';
        this.config = config;
        this.class = { Time, Sample };
        this.timeObject = new Time(this);
      }

      exit() {
        this.timeObject = undefined;
        delete window.ServerClock;
        console.log('Server Clock finished.');
      }

      synchronize() {
        return this.timeObject.synchronize();
      }

      getClientTime() {
        return this.class.Time.getClientTime();
      }

      getServerTime() {
        return this.timeObject.getServerTime();
      }
    };

(() => {
  let serverclock = defined ? window.ServerClock : new ServerClock();
  if (!window.location.href.includes('timeapi.io')) {
    serverclock.exit();
    alert('Go to https://www.timeapi.io/ and try again!');
    return;
  }
  if (!defined) {
    console.log(`Server Clock Test started. [Version ${serverclock.version}]`);
    window.ServerClock = serverclock;
  }
  serverclock.synchronize();
})();


js/serverClock.js
const config = {
  sampleMinimum: 3,
  sampleMaximum: 10,
  timeoutAfter: 5000, // In msec. Will retry request after this time
  validFor: 1200, // In sec. Will assume to be inaccurate after x seconds
  errorTolerance: 100,
  targetURL: window.location.href,
};
const defined = typeof window.ServerClocks !== 'undefined';
let Time = defined
  ? window.ServerClocks[0].class.Time
  : class Time {
      constructor(SC) {
        this.SC = SC;
        this.config = SC.config;
        this.currentAdjustment = 0; // The chosen adjustment
        this.lastSynchronized = 0;
      }

      static getClientTime() {
        return performance.timeOrigin + performance.now();
      }

      getServerTime() {
        return Time.getClientTime() + this.currentAdjustment;
      }

      synchronize() {
        this.SC.stop = false;
        this.lastSynchronized = performance.now();
        let differenceSamples = []; // A pool of potential clock adjustments (client/server time difference)
        let requestCount = 0;
        let expectedElapseAfterRequest = 0;
        this.SC.uiObject.updateMessage('3cb371', 'SYNCHRONIZING...');

        return new Promise((resolve) => {
          const recursiveSynchronization = async (resolver) => {
            if (this.SC.stop) return;
            try {
              requestCount++;
              const sample = await new Sample(this.SC).collect();
              differenceSamples.push(sample.HTTPAdjustment);
              differenceSamples.sort((a, b) => b - a);
              this.currentAdjustment = differenceSamples[0];
              expectedElapseAfterRequest = sample.elapsedSinceRequest;

              const nextStep = this.determineNextStep(
                requestCount,
                this.currentAdjustment,
                differenceSamples[differenceSamples.length - 1],
              );

              switch (nextStep) {
                case 2:
                  differenceSamples = differenceSamples.slice(1, -1);
                case 1:
                  setTimeout(
                    () => recursiveSynchronization(resolver),
                    this.getDelay(expectedElapseAfterRequest, differenceSamples.length),
                  );
                  break;
                case 0:
                  this.chooseAdjustment(differenceSamples);
                  resolver();
              }
            } catch (error) {
              if (requestCount < this.config.sampleMaximum || differenceSamples.length === 0) {
                setTimeout(
                  () => recursiveSynchronization(resolver),
                  this.getDelay(expectedElapseAfterRequest, differenceSamples.length),
                );
              } else {
                console.warn('Inaccuracy Warning: Request limit reached.');
                this.SC.uiObject.updateMessage('ffca28', 'INACCURACY WARNING');
                this.chooseAdjustment(differenceSamples);
              }
            }
          };
          recursiveSynchronization(resolve);
        });
      }

      determineNextStep(requestCount, max, min) {
        if (requestCount < this.config.sampleMinimum) return 1;
        if (requestCount >= this.config.sampleMaximum) {
          console.warn('Inaccuracy Warning: Request limit reached.');
          this.SC.uiObject.updateMessage('ffca28', 'INACCURACY WARNING');
          return 0;
        }
        if (max - min > 1000) return 2; // Likely outlier

        // Return whether the maximum value and the minimum value is around 1 sec (max truncation difference)
        return 1000 - this.config.errorTolerance > max - min ? 1 : 0;
      }

      getDelay(expectedElapseAfterRequest, precisionLevel) {
        const precision = 1000 / Math.pow(2, precisionLevel);
        const sampleOffset = Math.max(precision, this.config.errorTolerance);
        const adjustedTime = Time.getClientTime() + this.currentAdjustment;
        const delay = 1000 - ((adjustedTime + expectedElapseAfterRequest + sampleOffset) % 1000);

        return delay;
      }

      chooseAdjustment(differenceSamples) {
        const candidates = [];
        for (let i = 0; i < differenceSamples.length; i++) {
          for (let j = differenceSamples.length - 1; j >= 0; j--) {
            if (differenceSamples[i] - differenceSamples[j] < 1000) {
              candidates.push([j - i, i]);
              break;
            }
          }
        }
        candidates.sort((a, b) => b[0] - a[0]);
        this.currentAdjustment = differenceSamples[candidates[0][1]];

        console.log('[Adjustments]');
        differenceSamples.forEach((element) => {
          console.log(element / 1000);
        });
        console.log('The chosen adjustment is', this.currentAdjustment / 1000, 'sec.');
        if (differenceSamples.length < this.config.sampleMaximum) this.SC.uiObject.updateMessage('000', '');
        this.lastSynchronized = performance.now();
      }
    };

let Sample = defined
  ? window.ServerClocks[0].class.Sample
  : class Sample {
      constructor(SC) {
        this.SC = SC;
        this.config = SC.config;
        this.timeIdErrorTolerance = 500;
        this.controller = new AbortController();
        this.observer;
        this.requestTimeout;
        this.request = new Request(this.config.targetURL, {
          // Create a request object with no-cache headers
          // This prevents looking up 'date' header from cached responses
          method: 'GET',
          cache: 'no-store',
          headers: new Headers({
            Connection: 'keep-alive',
            'Cache-Control': 'no-cache',
            Pragma: 'no-cache',
            Expires: '0',
          }),
        });
      }

      collect() {
        return new Promise((resolve, reject) => {
          let HTTPTime = 0;
          let timeId = 0;
          this.observer = new PerformanceObserver((list) => {
            const entry = list.getEntries().find(({ name }) => name === this.request.url);
            if (entry && Math.abs(timeId - entry.startTime) < this.timeIdErrorTolerance) {
              clearTimeout(this.requestTimeout);
              this.observer.disconnect();
              this.processSample(resolve, entry, requestStartTime, HTTPTime);
            } else {
              this.setTimeout(reject);
            }
          });
          this.observer.observe({ type: 'resource' });

          timeId = performance.now();
          const requestStartTime = Time.getClientTime();
          fetch(this.request, { signal: this.controller.signal })
            .then(async (response) => {
              HTTPTime = Date.parse(response.headers.get('date'));
              await response.text(); // Fixes incomplete response issue
            })
            .catch(() => {});

          this.setTimeout(reject);
        });
      }

      processSample(returnSuccess, entry, requestStartTime, HTTPTime) {
        const estimatedLatency = (entry.responseStart - entry.requestStart) / 2;
        const elapsedSinceRequest = entry.requestStart - entry.startTime + estimatedLatency;

        // Calculate client/server time difference based on HTTP `date` header
        // Accomodate estimated elapsed time (time taken before server recorded `date`)
        const HTTPAdjustment = HTTPTime - (requestStartTime + elapsedSinceRequest);

        const result = {
          elapsedSinceRequest,
          HTTPAdjustment,
        };

        console.log('Collected a difference sample.');
        returnSuccess(result);
      }

      setTimeout(returnFail) {
        if (this.requestTimeout) clearTimeout(this.requestTimeout);
        this.requestTimeout = setTimeout(() => {
          console.log('Request timed out. Timeout:', config.timeoutAfter / 1000, 'sec.');
          this.observer.disconnect();
          this.controller.abort();
          returnFail();
        }, this.config.timeoutAfter);
      }
    };

let UI = defined
  ? window.ServerClocks[0].class.UI
  : class UI {
      static getSeed = () => Math.floor(Math.random() * (2 << 25));
      static getName = (seed, offset) => `e${(seed + offset).toString(16)}`;
      static formatTime = (clockTime) => {
        const formatDigit = (d) => String(d).padStart(2, '0');
        const time = new Date(clockTime);
        const hours = formatDigit(time.getHours() % 12 || 12);
        const minutes = formatDigit(time.getMinutes());
        const seconds = formatDigit(time.getSeconds());
        const amPm = time.getHours() >= 12 ? 'PM' : 'AM';
        return `${hours}:${minutes}:${seconds} ${amPm}`;
      };

      constructor(SC) {
        this.SC = SC;
        this.config = SC.config;
        this.updateTimer;
        this.clockElements;
        this.elementID = this.makeElementId();
        this.clockElements = this.createClockElements();
        this.styleSheet = document.createElement('style');
        this.constructUI();
      }

      makeElementId() {
        const elementID = {};
        const nameSeed = UI.getSeed();
        elementID.clock = UI.getName(nameSeed, 0);
        elementID.time = UI.getName(nameSeed, 1);
        elementID.message = UI.getName(nameSeed, 2);
        elementID.contextmenu = UI.getName(nameSeed, 3);
        elementID.menuitem = UI.getName(nameSeed, 4);
        return elementID;
      }

      createClockElements() {
        const frameElement = document.createElement('div');
        frameElement.id = this.elementID.clock;
        frameElement.draggable = true;
        let initialX, initialY;

        const timeElement = document.createElement('span');
        timeElement.textContent = '00:00:00 NA';
        timeElement.id = this.elementID.time;
        frameElement.appendChild(timeElement);

        const messageElement = document.createElement('span');
        messageElement.id = this.elementID.message;
        frameElement.appendChild(messageElement);

        frameElement.addEventListener('dragstart', (event) => {
          initialX = event.clientX - frameElement.offsetLeft;
          initialY = event.clientY - frameElement.offsetTop;
        });

        frameElement.addEventListener('dragend', (event) => {
          const newX = event.clientX - initialX;
          const newY = event.clientY - initialY;

          const borderX = window.innerWidth - frameElement.offsetWidth;
          const borderY = window.innerHeight - frameElement.offsetHeight;

          const applyX = Math.max(0, Math.min(newX, borderX));
          const applyY = Math.max(0, Math.min(newY, borderY));

          frameElement.style.left = `${applyX}px`;
          frameElement.style.top = `${applyY}px`;
        });

        return [frameElement, timeElement, messageElement];
      }

      attachContextMenu(element, menuItems) {
        element.addEventListener('contextmenu', (event) => {
          event.preventDefault();

          if (document.getElementById(this.elementID.contextmenu)) {
            const menu = document.getElementById(this.elementID.contextmenu);
            menu.style.top = `${event.clientY}px`;
            menu.style.left = `${event.clientX}px`;
            return;
          }

          const menu = document.createElement('div');
          menu.id = this.elementID.contextmenu;
          menu.style.top = `${event.clientY}px`;
          menu.style.left = `${event.clientX}px`;

          for (const item of menuItems) {
            const menuItem = document.createElement('div');
            menuItem.textContent = item.label;
            menuItem.addEventListener('click', item.action);
            menuItem.classList.add(this.elementID.menuitem);
            menu.appendChild(menuItem);
          }

          const closeMenu = () => {
            document.body.removeChild(menu);
            document.removeEventListener('click', closeMenu);
            document.removeEventListener('dragstart', closeMenu);
          };

          document.body.appendChild(menu);
          document.addEventListener('click', closeMenu);
          document.addEventListener('dragstart', closeMenu);
        });
      }

      action_rerun() {
        console.log('Resynchronizing...');
        this.SC.timeObject.synchronize().then(this.startUpdateTimer());
      }

      action_visitproject() {
        console.log('Opening the project page...');
        window.open('https://github.com/r8btx/Server-Clock-Bookmarklet/', '_blank');
      }

      action_exit() {
        this.SC.exit();
      }

      removeUI() {
        if (typeof this.updateTimer !== 'undefined') clearTimeout(this.updateTimer);
        document.body.removeChild(this.clockElements[0]);
        document.head.removeChild(this.styleSheet);
      }

      constructUI() {
        const elementID = this.elementID;
        const clockElements = this.clockElements;
        const styleSheet = this.styleSheet;
        const menuItems = [
          { label: 'Rerun', action: () => this.action_rerun() },
          { label: 'Visit project page', action: () => this.action_visitproject() },
          { label: 'Exit', action: () => this.action_exit() },
        ];

        this.attachContextMenu(clockElements[0], menuItems);

        const highestZIndex = Array.from(document.querySelectorAll('*')).reduce((maxZIndex, element) => {
          const zIndex = parseInt(getComputedStyle(element).zIndex);
          return isNaN(zIndex) ? maxZIndex : Math.max(maxZIndex, zIndex);
        }, 0);

        const styles = `
          #${elementID.clock}:hover,
          #${elementID.contextmenu} {
            box-shadow: 0 2px 5px rgba(0, 0, 0, 0.2);
          }
          #${elementID.clock},
          #${elementID.contextmenu} {
            cursor: pointer;
            user-select: none;
            display: flex;
            flex-direction: column;
            border-radius: 4px;
            font-family: Arial, sans-serif;
            position: fixed;
            left: 10px;
            background: #fff;
            color: #000;
            z-index: ${highestZIndex};
          }
          #${elementID.clock} {
            top: 10px;
            padding: 12px;
            border: 1px solid #ddd;
            border-radius: 6px;
            align-items: center;
            transition: box-shadow 0.3s;
          }
          #${elementID.message} {
            line-height: 0;
            font-size: 0.6rem;
            font-weight: 700;
          }
          #${elementID.time} {
            font-size: 1.25rem;
            line-height: 1.75;
          }
          #${elementID.contextmenu} {
            padding: 6px 0;
          }
          .${elementID.menuitem} {
            font-family: Arial, sans-serif;
            font-size: 1.125rem;
            line-height: 1.75rem;
            padding: 6px 12px;
            cursor: pointer;
            transition: background-color 0.3s;
          }
          .${elementID.menuitem}:hover {
            background-color: #3cb371;
            color: #fff;
          }`;

        styleSheet.innerText = styles.replace(/\n/g, '').replace(/\s+/g, ' ');
        document.head.appendChild(styleSheet);

        document.body.appendChild(clockElements[0]);
      }

      updateClock() {
        if ((performance.now() - this.SC.timeObject.lastSynchronized) / 1000 > this.config.validFor)
          this.updateMessage('ffca28', 'INACCURACY WARNING');

        const timeElement = this.clockElements[1];
        const clockTime = this.SC.timeObject.getServerTime();
        timeElement.textContent = UI.formatTime(clockTime);

        this.updateTimer = this.SC.stop
          ? this.updateMessage('ff5252', 'STOPPED')
          : setTimeout(this.updateClock.bind(this), 1000 - (clockTime % 1000));
      }

      startUpdateTimer() {
        if (typeof this.updateTimer === 'undefined' && typeof this.SC.timeObject !== 'undefined') {
          this.SC.stop = false;
          this.updateClock();
        }
      }

      updateMessage = (hexColor, message) => {
        const messageDOM = document.getElementById(this.elementID.message);
        if (!messageDOM) return;
        messageDOM.textContent = message;
        messageDOM.style.color = `#${hexColor}`;
      };
    };

class ServerClock {
  constructor() {
    this.version = '1.4';
    this.config = config;
    this.stop = false;
    this.class = { Time, Sample, UI };
    this.timeObject = new Time(this);
    this.uiObject = new UI(this);
  }

  exit() {
    this.stop = true;
    this.uiObject.removeUI();
    const index = window.ServerClocks.indexOf(this);
    window.ServerClocks.splice(index, 1);
    if (window.ServerClocks.length === 0) delete window.ServerClocks;

    this.timeObject = undefined;
    this.uiObject = undefined;
    console.log('Server Clock finished.');
  }

  synchronize() {
    return this.timeObject.synchronize();
  }

  getClientTime() {
    return this.class.Time.getClientTime();
  }

  getServerTime() {
    return this.timeObject.getServerTime();
  }

  startUpdateTimer() {
    this.uiObject.startUpdateTimer();
  }
}

(() => {
  let serverclock = new ServerClock();
  console.log(`Server Clock started. [Version ${serverclock.version}]`);
  serverclock.synchronize().then(() => serverclock.startUpdateTimer());
  if (!defined) {
    window.ServerClocks = [];
  }
  window.ServerClocks.push(serverclock);
})();
